src/accumulator/proof.rs

```diff
@@ -241,6 +243,286 @@ impl Proof {
+        for pos in needed_proof_positions {
+            for (new_node_pos, hash) in new_nodes.iter() {
+                if *new_node_pos == pos {
+                    new_proof.push((pos, *hash));
+                }
+            }
+        }
+        new_proof.sort();
+        let (_, hashes): (Vec<u64>, Vec<sha256::Hash>) = new_proof.into_iter().unzip();
+        Ok((
+            Proof {
+                hashes,
+                targets: new_target_pos,
+            },
+            target_hashes,
+        ))
+    }
+    /// maybe_remap remaps the passed in hash and pos if the treeRows increase after
+    /// adding the new nodes.
+    fn maybe_remap(
+        num_leaves: u64,
+        num_adds: u64,
+        positions: Vec<(u64, sha256::Hash)>,
+    ) -> Vec<(u64, sha256::Hash)> {
+        let new_forest_rows = util::tree_rows(num_leaves + num_adds);
+        let old_forest_rows = util::tree_rows(num_leaves);
+        let tree_rows = util::tree_rows(num_leaves);
+        let mut new_proofs = vec![];
+        if new_forest_rows > old_forest_rows {
+            for (pos, hash) in positions.iter() {
+                let row = util::detect_row(*pos, tree_rows);
+
+                let old_start_pos = util::start_position_at_row(row, old_forest_rows);
+                let new_start_pos = util::start_position_at_row(row, new_forest_rows);
+
+                let offset = pos - old_start_pos;
+                let new_pos = offset + new_start_pos;
+                new_proofs.push((new_pos, *hash));
+            }
+            return new_proofs;
+        }
+
+        positions
+    }
+
+    /// update_proof_remove modifies the cached proof with the deletions that happen in the block proof.
+    /// It updates the necessary proof hashes and un-caches the targets that are being deleted.
+    fn update_proof_remove(
+        self,
+        block_targets: Vec<u64>,
+        cached_hashes: Vec<sha256::Hash>,
+        updated: Vec<(u64, sha256::Hash)>,
+        num_leaves: u64,
+    ) -> Result<(Proof, Vec<sha256::Hash>), String> {
+        let total_rows = util::tree_rows(num_leaves);
+
+        let targets_with_hash: Vec<(u64, bitcoin_hashes::sha256::Hash)> = self
+            .targets
+            .clone()
+            .into_iter()
+            .zip(cached_hashes.clone().into_iter())
+            .filter(|(pos, _)| !block_targets.contains(pos))
+            .collect();
+        let (targets, target_hashes) = targets_with_hash.to_owned().into_iter().unzip();
+
+        let proof_positions =
+            util::get_proof_positions(&self.targets, num_leaves, util::tree_rows(num_leaves));
+
+        let old_proof: Vec<_> = proof_positions
+            .iter()
+            .zip(self.hashes.clone().into_iter())
+            .collect();
+
+        let mut new_proof = vec![];
+        // Grab all the positions of the needed proof hashes.
+        let needed_pos = util::get_proof_positions(&targets, num_leaves, total_rows);
+
+        let updated: Vec<_> = updated
```

Comment:
We should be creating a `extra_positions` variable instead of shadowing `updated` since we'll need it for adding new proofs.

Take for example this tree:
```
14
|---------------\
12              13
|-------\       |-------\
08      09      10      11
|---\   |---\   |---\   |---\
00  01  02  03  04  05  06  07
```

If we cache `[00, 07]`, the hashes of `12` and `13` are left out as they're able to be calculated from hashing either `(00 | 01) | 09` or `10 | (06 | 07)`. But if we delete `0`, then `12` isn't able to be calculated.

`updated` provides the new hash for `12` after the deletion of `0` and we use this data to add in the missing `12`. But the code here filters out `12`, so this ends up creating an invalid proof.

Here's the scenario I described above as a test case:

```
    {
        "update": {
            "adds": [],
            "proof": {
                "targets": [0],
                "hashes": [
                    "4bf5122f344554c53bde2ebb8cd2b7e3d1600ad631c385a5d7cce23c7785459a",
                    "9576f4ade6e9bc3a6458b506ce3e4e890df29cb14cb5d3d887672aef55647a2b",
                    "29590a14c1b09384b94a2c0e94bf821ca75b62eacebc47893397ca88e3bbcbd7"
                ]
            },
            "del_hashes": [
                "6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d"
            ]
        },
        "remembers": [],
        "cached_proof": {
            "targets": [0, 7],
            "hashes": [
                "4bf5122f344554c53bde2ebb8cd2b7e3d1600ad631c385a5d7cce23c7785459a",
                "67586e98fad27da0b9968bc039a1ef34c939b9b8e523a8bef89d478608c5ecf6",
                "9576f4ade6e9bc3a6458b506ce3e4e890df29cb14cb5d3d887672aef55647a2b",
                "9eec588c41d87b16b0ee226cb38da3864f9537632321d8be855a73d5616dcc73"
            ]
        },
        "initial_stump": "Add 8 leaves [0, 1, 2, 3, 4, 5, 6, 7]",
        "initial_roots": [
            "b151a956139bb821d4effa34ea95c17560e0135d1e4661fc23cedc3af49dac42"
        ],
        "initial_leaves": 8,
        "cached_hashes": [
            "6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d",
            "ca358758f6d27e6cf45272937977a748fd88391db679ceda7dc7bf1f005ee879"
        ],
        "expected_roots": [
            "726fdd3b432cc59e68487d126e70f0db74a236267f8daeae30b31839a4e7ebed"
        ],
        "expected_targets": [7],
        "expected_cached_hashes": [
            "ca358758f6d27e6cf45272937977a748fd88391db679ceda7dc7bf1f005ee879"
        ]
    }
```

Reply:
I changed how we deal with old proof, now we use `updated` without filtering, but only asks for positions that are required for the new proof. Because using `updated` as passed in would create nodes that are no longer needed. In this case, `01` and `09`. The filtered version is not used anymore.